### PR TITLE
WIP: Difficulty: new class representing the difficulty target of a block

### PR DESCRIPTION
**Pull request by schildbach**
TODO add back deprecated methods
TODO test more edge cases
TODO check bounds of `of*()` static constructors, but what are they?

### CODE REVIEW

**Comment by schildbach** [comment]
This PR is now ready for review.

**Comment by schildbach** [comment]
Rebased.

**Comment by schildbach** [comment]
I added a couple of small improvements on top.

**Comment by msgilligan** [comment]
Looks good after a quick glance. This is a candidate for a value-based class (Valhalla) -- we should take a look at the latest spec (which may make "preview" in JDK 25) and make any minor adjustments if needed.

**Comment by schildbach** [comment]
I assume we could also use int for the compact format. After all, it's only 4 bytes and the first byte is always between 0x00 and 0x20, inclusive. So there would not be a problem with the sign bit.

However, since this was a long before at least for this PR I'd like to stay with long, and refactor it in a future PR if so desired.

File: base/src/main/java/org/bitcoinj/base/Difficulty.java [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+ * The target difficulty is a value which a header hash must be equal to or below in order for that header to be a
+ * valid part of the block chain.
+ * <p>
+ * Difficulty values cannot be negative. A high value means low difficulty, and a low value means high difficulty.
+ * <p>
+ * The value is stored in only 32 bits of space, so it uses a less precise “compact form”, or "nBits". Think of it as a
+ * base-256 version of the scientific notation, consisting of a 1 byte exponent and 3 bytes of mantissa. That form is
+ * used for storing the value in this class as well, so when constructing from an integer value, be prepared to lose
+ * precision.
+ * <p>
```
Comment (msgilligan) [inline base/src/main/java/org/bitcoinj/base/Difficulty.java:37]: Could we provide a few examples of compact format here?  (Alternatively we could link to some of the static constants as examples)

File: base/src/main/java/org/bitcoinj/base/Difficulty.java [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+        checkArgument(!actualTimespan.isZero(), () -> "cannot be zero: " + actualTimespan);
+        checkArgument(!targetTimespan.isNegative(), () -> "cannot be negative: " + targetTimespan);
+        checkArgument(!targetTimespan.isZero(), () -> "cannot be zero: " + targetTimespan);
+        BigInteger adjusted = asInteger()
+                .multiply(BigInteger.valueOf(actualTimespan.getSeconds()))
+                .divide(BigInteger.valueOf(targetTimespan.getSeconds()));
+        if (adjusted.compareTo(maxTarget.asInteger()) > 0)
+            return maxTarget;
+        else
+            return Difficulty.ofInteger(adjusted);
```
Comment (msgilligan) [inline base/src/main/java/org/bitcoinj/base/Difficulty.java:134]: I recommend a single `return` statement here with a ternary operator.

File: base/src/main/java/org/bitcoinj/base/Difficulty.java [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+     * @return adjusted difficulty to be used for the next interval
+     */
+    public Difficulty adjust(Duration actualTimespan, Duration targetTimespan, Difficulty maxTarget) {
+        checkArgument(!actualTimespan.isNegative(), () -> "cannot be negative: " + actualTimespan);
+        checkArgument(!actualTimespan.isZero(), () -> "cannot be zero: " + actualTimespan);
+        checkArgument(!targetTimespan.isNegative(), () -> "cannot be negative: " + targetTimespan);
+        checkArgument(!targetTimespan.isZero(), () -> "cannot be zero: " + targetTimespan);
+        BigInteger adjusted = asInteger()
+                .multiply(BigInteger.valueOf(actualTimespan.getSeconds()))
+                .divide(BigInteger.valueOf(targetTimespan.getSeconds()));
```
Comment (msgilligan) [inline base/src/main/java/org/bitcoinj/base/Difficulty.java:130]: This might read a little easier if you create two intermediate `BigInteger` variables for "actual" and "target" and then do the divide/multiply with them. Your call.

**Review by msgilligan — CHANGES_REQUESTED** [review]
This looks great! I have a few minor, nit-picky suggestions.

File: base/src/main/java/org/bitcoinj/base/Difficulty.java [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+import org.bitcoinj.base.internal.ByteUtils;
+
+import java.math.BigInteger;
+import java.time.Duration;
+import java.util.Objects;
+
+import static org.bitcoinj.base.internal.Preconditions.checkArgument;
+
+/**
+ * The target difficulty is a value which a header hash must be equal to or below in order for that header to be a
```
[the remark this answers is not in the document]
Comment (schildbach) [reply base/src/main/java/org/bitcoinj/base/Difficulty.java]: Fixed via 8746b447be232512e36ab169e0a33edf5d8b041b.

File: base/src/main/java/org/bitcoinj/base/Difficulty.java (hunk 2 above)
Comment (schildbach) [reply base/src/main/java/org/bitcoinj/base/Difficulty.java]: Done: af278ed51f9fba4ab86338f01113f067b347edf9

File: base/src/main/java/org/bitcoinj/base/Difficulty.java (hunk 3 above)
Comment (schildbach) [reply base/src/main/java/org/bitcoinj/base/Difficulty.java]: Done: be2256d27

File: base/src/main/java/org/bitcoinj/base/Difficulty.java (hunk 1 above)
Comment (schildbach) [reply base/src/main/java/org/bitcoinj/base/Difficulty.java:37]: The best explanation I've seen is https://learnmeabitcoin.com/technical/block/bits/, but do we want to link to such a site?

Comment (schildbach) [reply base/src/main/java/org/bitcoinj/base/Difficulty.java:37]: But I added a link to one of the constants: 90aa14c67

**Comment by schildbach** [comment]
Requesting re-review to see if you're ok with the added link to an example compact format.

File: base/src/main/java/org/bitcoinj/base/Difficulty.java (hunk 1 above)
Comment (msgilligan) [reply base/src/main/java/org/bitcoinj/base/Difficulty.java:37]: > do we want to link to such a site?

Probably not since it is "unofficial" and could disappear or get squatted.

I think the added link is good enough.

**Review by msgilligan — APPROVED** [review]
Approved!
